storage/clickhouse/clickhouse.go

```diff
@@ -297,26 +349,43 @@ func (ch *clickHouse) Read(ctx context.Context, query *prompb.Query) (*prompb.Qu
 
 	convertedQuery := ch.convertReadRequest(query)
 
-	res := new(prompb.QueryResult)
-	// find matching time series
-	fingerprints := make(map[uint64]struct{}, 64)
-	ch.timeSeriesRW.RLock()
-	for f, labels := range ch.timeSeries {
-		if convertedQuery.Matchers.MatchLabels(labels) {
-			fingerprints[f] = struct{}{}
-		}
-	}
-	ch.timeSeriesRW.RUnlock()
-	if len(fingerprints) == 0 {
-		return res, nil
-	}
 	var metricName string
 	for _, matcher := range convertedQuery.Matchers {
 		if matcher.Name == "__name__" {
 			metricName = matcher.Value
 		}
 	}
 
+	var clickHouseQuery string
+	var fingerprints map[uint64]struct{}
+	var err error
+
+	if ch.useClickHouseQuery {
+		clickHouseQuery, err = ch.prepareClickHouseQuery(query, metricName)
+		if err != nil {
+			return nil, err
+		}
+		fingerprints, err = ch.fingerprintsForQuery(ctx, clickHouseQuery)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		fingerprints = make(map[uint64]struct{}, 64)
+		// find matching time series
+		ch.timeSeriesRW.RLock()
+		for f, labels := range ch.timeSeries {
```

Comment:
If we can use `metric_name` column here, we could avoid matching many unnecessary fingerprints, right?

Reply:
Can you elaborate? The `MatchLabels` function already does an optimization around the metric name. It breaks out of the loop early based on the special `__name__` key.